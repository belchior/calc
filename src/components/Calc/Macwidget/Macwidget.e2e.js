/* eslint jest/valid-expect-in-promise: 0 */
describe ('Calc', function () {
  browser.ignoreSynchronization = true;
  browser.get('http://localhost:3000/');

  it('click on menu item Macwidget', function () {
    $('.link-macwidget').click();
  });

  describe('Macwidget', function () {

    var macwidget = {};
    macwidget.number0 = $('.Macwidget .btn[data-name=number0]');
    macwidget.number1 = $('.Macwidget .btn[data-name=number1]');
    macwidget.number2 = $('.Macwidget .btn[data-name=number2]');
    macwidget.number3 = $('.Macwidget .btn[data-name=number3]');
    macwidget.number4 = $('.Macwidget .btn[data-name=number4]');
    macwidget.number5 = $('.Macwidget .btn[data-name=number5]');
    macwidget.number6 = $('.Macwidget .btn[data-name=number6]');
    macwidget.number7 = $('.Macwidget .btn[data-name=number7]');
    macwidget.number8 = $('.Macwidget .btn[data-name=number8]');
    macwidget.number9 = $('.Macwidget .btn[data-name=number9]');
    macwidget.addition = $('.Macwidget .btn[data-name=addition]');
    macwidget.subtraction = $('.Macwidget .btn[data-name=subtraction]');
    macwidget.multiplication = $('.Macwidget .btn[data-name=multiplication]');
    macwidget.division = $('.Macwidget .btn[data-name=division]');
    macwidget.equality = $('.Macwidget .btn[data-name=equality]');
    macwidget.dot = $('.Macwidget .btn[data-name=dot]');
    macwidget.clear = $('.Macwidget .btn[data-name=clear]');
    macwidget.memoryAdd = $('.Macwidget .btn[data-name=madd]');
    macwidget.memorySubtract = $('.Macwidget .btn[data-name=msubtract]');
    macwidget.memoryClear = $('.Macwidget .btn[data-name=mclear]');
    macwidget.memoryRecall = $('.Macwidget .btn[data-name=mrecall]');
    macwidget.display = $('.Macwidget .display');

    beforeEach(function () {
      macwidget.clear.click();
    });

    describe('Click on button', function () {
      it('Number 0', function () {
        macwidget.number0.click()
        .then(function () { return macwidget.display.getText(); })
        .then(function (text) { expect(text).toEqual('0'); });
      });
      it('Number 1', function () {
        macwidget.number1.click()
        .then(function () { return macwidget.display.getText(); })
        .then(function (text) { expect(text).toEqual('1'); });
      });
      it('Number 2', function () {
        macwidget.number2.click()
        .then(function () { return macwidget.display.getText(); })
        .then(function (text) { expect(text).toEqual('2'); });
      });
      it('Number 3', function () {
        macwidget.number3.click()
        .then(function () { return macwidget.display.getText(); })
        .then(function (text) { expect(text).toEqual('3'); });
      });
      it('Number 4', function () {
        macwidget.number4.click()
        .then(function () { return macwidget.display.getText(); })
        .then(function (text) { expect(text).toEqual('4'); });
      });
      it('Number 5', function () {
        macwidget.number5.click()
        .then(function () { return macwidget.display.getText(); })
        .then(function (text) { expect(text).toEqual('5'); });
      });
      it('Number 6', function () {
        macwidget.number6.click()
        .then(function () { return macwidget.display.getText(); })
        .then(function (text) { expect(text).toEqual('6'); });
      });
      it('Number 7', function () {
        macwidget.number7.click()
        .then(function () { return macwidget.display.getText(); })
        .then(function (text) { expect(text).toEqual('7'); });
      });
      it('Number 8', function () {
        macwidget.number8.click()
        .then(function () { return macwidget.display.getText(); })
        .then(function (text) { expect(text).toEqual('8'); });
      });
      it('Number 9', function () {
        macwidget.number9.click()
        .then(function () { return macwidget.display.getText(); })
        .then(function (text) { expect(text).toEqual('9'); });
      });

      // @TODO this test do not represent an aplication rule
      // it('Operator +', function () {
      //   macwidget.number1.click()
      //   .then(function () { macwidget.addition.click(); })
      //   .then(function () { macwidget.number2.click(); })
      //   .then(function () { macwidget.addition.click(); })
      //   .then(function () { return macwidget.display.getText(); })
      //   .then(function (text) { expect(text).toEqual('3'); });
      // });

      // @TODO this test do not represent an aplication rule
      // it('Operator -', function () {
      //   macwidget.number3.click()
      //   .then(function () { macwidget.subtraction.click(); })
      //   .then(function () { macwidget.number4.click(); })
      //   .then(function () { macwidget.subtraction.click(); })
      //   .then(function () { return macwidget.display.getText(); })
      //   .then(function (text) { expect(text).toEqual('-1'); });
      // });

      // @TODO this test do not represent an aplication rule
      // it('Operator ร', function () {
      //   macwidget.number5.click()
      //   .then(function () { macwidget.multiplication.click(); })
      //   .then(function () { macwidget.number6.click(); })
      //   .then(function () { macwidget.multiplication.click(); })
      //   .then(function () { return macwidget.display.getText(); })
      //   .then(function (text) { expect(text).toEqual('30'); });
      // });

      // @TODO this test do not represent an aplication rule
      // it('Operator รท', function () {
      //   macwidget.number7.click()
      //   .then(function () { macwidget.division.click(); })
      //   .then(function () { macwidget.number8.click(); })
      //   .then(function () { macwidget.division.click(); })
      //   .then(function () { return macwidget.display.getText(); })
      //   .then(function (text) { expect(text).toEqual('0.875'); });
      // });
      it('Equal', function () {
        macwidget.number9.click()
        .then(function () { macwidget.addition.click(); })
        .then(function () { macwidget.number1.click(); })
        .then(function () { macwidget.equality.click(); })
        .then(function () { return macwidget.display.getText(); })
        .then(function (text) { expect(text).toEqual('10'); });
      });
      it('Dot', function () {
        macwidget.number3.click()
        .then(function () { macwidget.dot.click(); })
        .then(function () { macwidget.number4.click(); })
        .then(function () { return macwidget.display.getText(); })
        .then(function (text) { expect(text).toEqual('3.4'); });
      });
      it('Clear', function () {
        macwidget.number1.click()
        .then(function () { macwidget.number2.click(); })
        .then(function () { macwidget.number3.click(); })
        .then(function () { macwidget.clear.click(); })
        .then(function () { return macwidget.display.getText(); })
        .then(function (text) { expect(text).toEqual(''); });
        // The original value are 0
        // .then(function (text) { expect(text).toEqual('0'); });
      });
      // it('Memory Recall', function () {
      //   macwidget.number4.click()
      //   .then(function () { macwidget.memoryRecall.click(); })
      //   .then(function () { return macwidget.display.getText(); })
      //   .then(function (text) { expect(text).toEqual('0'); });
      // });
      // it('Memory Add', function () {
      //   macwidget.number5.click()
      //   .then(function () { macwidget.memoryAdd.click(); })
      //   .then(function () { macwidget.memoryAdd.click(); })
      //   .then(function () { macwidget.memoryRecall.click(); })
      //   .then(function () { return macwidget.display.getText(); })
      //   .then(function (text) { expect(text).toEqual('10'); });
      // });
      // it('Memory Subtract', function () {
      //   macwidget.number5.click()
      //   .then(function () { macwidget.memorySubtract.click(); })
      //   .then(function () { macwidget.memoryRecall.click(); })
      //   .then(function () { return macwidget.display.getText(); })
      //   .then(function (text) { expect(text).toEqual('5'); })
      //
      //   .then(function () { macwidget.number3.click(); })
      //   .then(function () { macwidget.memorySubtract.click(); })
      //   .then(function () { macwidget.memoryRecall.click(); })
      //   .then(function () { return macwidget.display.getText(); })
      //   .then(function (text) { expect(text).toEqual('2'); });
      // });
    });

  });
});
