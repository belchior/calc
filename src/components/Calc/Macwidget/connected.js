import { connect } from 'react-redux';
import Macwidget from './Macwidget';

export const mapStateToProps = (state) => {
  return state.macwidget;
};

export const mapDispatchToProps = (dispatch) => {
  return {
    clearClick(e) {
      dispatch({ type: 'CLEAR', calc: 'macwidget', payload: e.target.getAttribute('data-value') });
    },
    divisionClick(e) {
      dispatch({ type: 'DIVISION', calc: 'macwidget', payload: e.target.getAttribute('data-value') });
    },
    dotClick(e) {
      dispatch({ type: 'DOT', calc: 'macwidget', payload: e.target.getAttribute('data-value') });
    },
    equalsClick(e) {
      dispatch({ type: 'EQUALS', calc: 'macwidget', payload: e.target.getAttribute('data-value') });
    },
    memoryClearClick() {
      dispatch({ type: 'MEMORY_CLEAR', calc: 'macwidget' });
    },
    memoryMinusClick() {
      dispatch({ type: 'MEMORY_MINUS', calc: 'macwidget' });
    },
    memoryPlusClick() {
      dispatch({ type: 'MEMORY_PLUS', calc: 'macwidget' });
    },
    memoryRecallClick() {
      dispatch({ type: 'MEMORY_RECALL', calc: 'macwidget' });
    },
    minusClick(e) {
      dispatch({ type: 'MINUS', calc: 'macwidget', payload: e.target.getAttribute('data-value') });
    },
    multiplicationClick(e) {
      dispatch({ type: 'MULTIPLY', calc: 'macwidget', payload: e.target.getAttribute('data-value') });
    },
    numberClick(e) {
      dispatch({ type: 'NUMBER', calc: 'macwidget', payload: e.target.getAttribute('data-value') });
    },
    plusClick(e) {
      dispatch({ type: 'PLUS', calc: 'macwidget', payload: e.target.getAttribute('data-value') });
    },
    disableError() {
      setTimeout(() => dispatch({ type: 'ERROR', calc: 'macwidget', payload: false }), 400);
    },
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(Macwidget);
