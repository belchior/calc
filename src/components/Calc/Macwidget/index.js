import Macwidget, {mapStateToProps, mapDispatchToProps} from './connected';

export {mapStateToProps, mapDispatchToProps};
export default Macwidget;
