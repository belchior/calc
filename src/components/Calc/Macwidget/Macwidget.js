import React, { Component } from 'react';
import Calc from '../Calc';
import './Macwidget.css';

class Macwidget extends Component {
  constructor(props) {
    super(props);
    this.state = {
      formula: '',
      error: false,
      startNewCalc: false,
      slot: 0
    };

    this.memory = {
      get: () => {
        return this.state.slot;
      },
      set: (value) => {
        this.setState({slot: value});
        return this;
      },
      sum: (value) => {
        this.setState({slot: Calc.sum(this.state.slot, value)});
        return this;
      },
      subtract: (value) => {
        this.setState({slot: Calc.subtract(this.state.slot, value)});
        return this;
      }
    };

    this.additionRule = this.additionRule.bind(this);
    this.clearRule = this.clearRule.bind(this);
    this.divisionRule = this.divisionRule.bind(this);
    this.dotRule = this.dotRule.bind(this);
    this.equalityRule = this.equalityRule.bind(this);
    this.maddRule = this.maddRule.bind(this);
    this.mclearRule = this.mclearRule.bind(this);
    this.mrecallRule = this.mrecallRule.bind(this);
    this.msubtractRule = this.msubtractRule.bind(this);
    this.multiplicationRule = this.multiplicationRule.bind(this);
    this.numberRule = this.numberRule.bind(this);
    this.subtractionRule = this.subtractionRule.bind(this);

    this.showError = this.showError.bind(this);
    this.buttonClick = this.buttonClick.bind(this);
  }

  additionRule(formula, char = '+') {
    if (formula && formula.slice(-1).search(/[+\-×÷]/) >= 0) {
      return formula.slice(0, -1) + char;
    }
    if (formula.slice(-1).search(/[.]/) < 0) {
      return formula + char;
    }
  }

  clearRule(formula) {
    this.setState({startNewCalc: true});
    return '';
  }

  divisionRule(formula, char = '÷') {
    if (formula && formula.slice(-1).search(/[+\-×÷]/) >= 0) {
      if (formula.length > 1 && formula.slice(-2).search(/[(^√]/) < 0) {
        return formula.slice(0, -1) + char;
      }

    } else if (formula && formula.slice(-1).search(/[(.^√]/) < 0) {
      return formula + char;
    }
  }

  dotRule(formula, char = '.') {
    if (
      formula && formula.slice(-1).search(/[%^√π.+\-×÷()]/) < 0 &&
      formula.search(/\d+\.\d+$/) < 0
    ) {
      return formula + char;
    }
  }

  equalityRule(formula) {
    if (formula === '') {
      return;
    }
    try {
      let result = String(Calc.calculate(formula));
      this.setState({startNewCalc: true});
      return result;

    } catch (err) {
      console.error(err);
    }
  }

  maddRule(formula) {
    if (formula.match(/^[+-]?\d+(?:\.\d+)?$/)) {
      this.memory.sum(parseFloat(formula));
      return false;
    }
  }

  mclearRule(formula) {
    this.memory.set(0);
    return '0';
  }

  mrecallRule(formula) {
    return this.memory.get().toString();
  }

  msubtractRule(formula) {
    if (formula.match(/^[+-]?\d+(?:\.\d+)?$/)) {
      this.memory.subtract(parseFloat(formula));
      return false;
    }
  }

  multiplicationRule(formula, char = '×') {
    if (formula && formula.slice(-1).search(/[+\-×÷]/) >= 0) {
      if (formula.length > 1 && formula.slice(-2).search(/[(^√]/) < 0) {
        return formula.slice(0, -1) + char;
      }

    } else if (formula && formula.slice(-1).search(/[(.^√]/) < 0) {
      return formula + char;
    }
  }

  numberRule(formula, char = '') {
    if (formula === '' || formula.slice(-1).search(/[)%π]/) < 0) {
      return formula + char;
    }
  }

  subtractionRule(formula, char = '-') {
    if (formula === '') {
      return formula + char;
    }
    if (formula.slice(-1).search(/[+\-×÷]/) >= 0) {
      return formula.slice(0, -1) + char;
    }
    if (formula.slice(-1).search(/[.]/) < 0) {
      return formula + char;
    }
  }

  showError() {
    this.setState({error: true});
    setTimeout(() => this.setState({error: false}), 400);
  }

  buttonClick(buttonRule) {
    return (e) => {
      const startNewCalc = this.state.startNewCalc;
      let formula = this.state.formula;

      if (startNewCalc) {
        formula = '';
        this.setState({startNewCalc: false});
      }
      formula = buttonRule(formula, e.target.dataset.value);

      if (formula === false) {
        return;
      }
      if (typeof formula !== 'string') {
        return this.showError();
      }
      this.setState({formula: formula});
    }
  }

  render() {
    return (
      <form className={this.state.error ? 'Macwidget shake-horizontal' : 'Macwidget'}>
        <output className="display">{this.state.formula}</output>
        <div className="keyboard">
          <div className="brackets">
            <button onClick={this.maddRule} type="button" className="btn small" data-name="madd" data-value="" title="memory add">m+</button>
            <button onClick={this.msubtractRule} type="button" className="btn small" data-name="msubtract" data-value="" title="memory subtract">m−</button>
            <button onClick={this.mclearRule} type="button" className="btn small" data-name="mclear" title="memory clear">mc</button>
            <button onClick={this.mrecallRule} type="button" className="btn small" data-name="mrecall" title="memory recall">mr</button>
            <button onClick={this.buttonClick(this.divisionRule)} type="button" className="btn small" data-name="division" data-value="÷" data-action="divide" title="divide">÷</button>
          </div>
          <div className="numbers">
            <button onClick={this.buttonClick(this.numberRule)} type="button" className="btn" data-name="number7" data-value="7" title="number 7">7</button>
            <button onClick={this.buttonClick(this.numberRule)} type="button" className="btn" data-name="number8" data-value="8" title="number 8">8</button>
            <button onClick={this.buttonClick(this.numberRule)} type="button" className="btn" data-name="number9" data-value="9" title="number 9">9</button>
            <button onClick={this.buttonClick(this.numberRule)} type="button" className="btn" data-name="number4" data-value="4" title="number 4">4</button>
            <button onClick={this.buttonClick(this.numberRule)} type="button" className="btn" data-name="number5" data-value="5" title="number 5">5</button>
            <button onClick={this.buttonClick(this.numberRule)} type="button" className="btn" data-name="number6" data-value="6" title="number 6">6</button>
            <button onClick={this.buttonClick(this.numberRule)} type="button" className="btn" data-name="number1" data-value="1" title="number 1">1</button>
            <button onClick={this.buttonClick(this.numberRule)} type="button" className="btn" data-name="number2" data-value="2" title="number 2">2</button>
            <button onClick={this.buttonClick(this.numberRule)} type="button" className="btn" data-name="number3" data-value="3" title="number 3">3</button>
            <button onClick={this.buttonClick(this.numberRule)} type="button" className="btn" data-name="number0" data-value="0" title="number 0">0</button>
            <button onClick={this.buttonClick(this.dotRule)} type="button" className="btn" data-name="dot" data-value="." title="dot">.</button>
            <button onClick={this.buttonClick(this.clearRule)} type="button" className="btn" data-name="clear" title="clear">c</button>
          </div>
          <div className="operators">
            <button onClick={this.buttonClick(this.multiplicationRule)} type="button" className="btn small" data-name="multiplication" data-value="×" data-action="multiply" title="multiply">×</button>
            <button onClick={this.buttonClick(this.subtractionRule)} type="button" className="btn small" data-name="subtraction" data-value="-" data-action="subtract" title="subtract">−</button>
            <button onClick={this.buttonClick(this.additionRule)} type="button" className="btn small" data-name="addition" data-value="+" data-action="sum" title="sum">+</button>
            <button onClick={this.buttonClick(this.equalityRule)} type="button" className="btn small" data-name="equality" title="equals">=</button>
          </div>
        </div>
      </form>
    );
  }
}

export default Macwidget;
