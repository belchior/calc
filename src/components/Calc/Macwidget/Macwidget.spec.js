import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import Macwidget from './Macwidget';


const setup = (props = {}) => {
  const requiredProps = {
    clearClick: () => {},
    disableError: () => {},
    divisionClick: () => {},
    dotClick: () => {},
    equalsClick: () => {},
    memoryPlusClick: () => {},
    memoryMinusClick: () => {},
    memoryClearClick: () => {},
    memoryRecallClick: () => {},
    minusClick: () => {},
    multiplicationClick: () => {},
    numberClick: () => {},
    plusClick: () => {},
    ...props,
  };
  return shallow(<Macwidget {...requiredProps} />);
};


describe('Macwidget', () => {
  it('should render without crashing', () => {
    const component = () => setup();
    expect(component).not.toThrow();
  });

  it('should add css class shake-horizontal when error props is equal true', () => {
    const props = {
      disableError: jest.fn(),
      error: true,
    };
    const wrapper = setup(props);
    const custom = toJson(wrapper);

    expect(props.disableError).toHaveBeenCalledTimes(1);
    expect(custom).toMatchSnapshot();
  });

  it('should render the value of formula into element with css class display', () => {
    const props = { formula: '456' };
    const wrapper = setup(props);
    const custom = toJson(wrapper);

    expect(custom).toMatchSnapshot();
  });
});
