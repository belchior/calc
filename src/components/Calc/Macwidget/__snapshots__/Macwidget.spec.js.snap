// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Macwidget should add css class shake-horizontal when error props is equal true 1`] = `
<form
  className="Macwidget shake-horizontal"
>
  <output
    className="display"
  />
  <div
    className="keyboard"
  >
    <div
      className="brackets"
    >
      <button
        className="btn small"
        data-name="madd"
        data-value=""
        title="memory add"
        type="button"
      >
        m+
      </button>
      <button
        className="btn small"
        data-name="msubtract"
        data-value=""
        title="memory subtract"
        type="button"
      >
        m−
      </button>
      <button
        className="btn small"
        data-name="mclear"
        title="memory clear"
        type="button"
      >
        mc
      </button>
      <button
        className="btn small"
        data-name="mrecall"
        title="memory recall"
        type="button"
      >
        mr
      </button>
      <button
        className="btn small"
        data-action="divide"
        data-name="division"
        data-value="÷"
        title="divide"
        type="button"
      >
        ÷
      </button>
    </div>
    <div
      className="numbers"
    >
      <button
        className="btn"
        data-name="number7"
        data-value="7"
        title="number 7"
        type="button"
      >
        7
      </button>
      <button
        className="btn"
        data-name="number8"
        data-value="8"
        title="number 8"
        type="button"
      >
        8
      </button>
      <button
        className="btn"
        data-name="number9"
        data-value="9"
        title="number 9"
        type="button"
      >
        9
      </button>
      <button
        className="btn"
        data-name="number4"
        data-value="4"
        title="number 4"
        type="button"
      >
        4
      </button>
      <button
        className="btn"
        data-name="number5"
        data-value="5"
        title="number 5"
        type="button"
      >
        5
      </button>
      <button
        className="btn"
        data-name="number6"
        data-value="6"
        title="number 6"
        type="button"
      >
        6
      </button>
      <button
        className="btn"
        data-name="number1"
        data-value="1"
        title="number 1"
        type="button"
      >
        1
      </button>
      <button
        className="btn"
        data-name="number2"
        data-value="2"
        title="number 2"
        type="button"
      >
        2
      </button>
      <button
        className="btn"
        data-name="number3"
        data-value="3"
        title="number 3"
        type="button"
      >
        3
      </button>
      <button
        className="btn"
        data-name="number0"
        data-value="0"
        title="number 0"
        type="button"
      >
        0
      </button>
      <button
        className="btn"
        data-name="dot"
        data-value="."
        title="dot"
        type="button"
      >
        .
      </button>
      <button
        className="btn"
        data-name="clear"
        title="clear"
        type="button"
      >
        c
      </button>
    </div>
    <div
      className="operators"
    >
      <button
        className="btn small"
        data-action="multiply"
        data-name="multiplication"
        data-value="×"
        title="multiply"
        type="button"
      >
        ×
      </button>
      <button
        className="btn small"
        data-action="subtract"
        data-name="subtraction"
        data-value="-"
        title="subtract"
        type="button"
      >
        −
      </button>
      <button
        className="btn small"
        data-action="sum"
        data-name="addition"
        data-value="+"
        title="sum"
        type="button"
      >
        +
      </button>
      <button
        className="btn small"
        data-name="equality"
        title="equals"
        type="button"
      >
        =
      </button>
    </div>
  </div>
</form>
`;

exports[`Macwidget should render the value of formula into element with css class display 1`] = `
<form
  className="Macwidget"
>
  <output
    className="display"
  >
    456
  </output>
  <div
    className="keyboard"
  >
    <div
      className="brackets"
    >
      <button
        className="btn small"
        data-name="madd"
        data-value=""
        title="memory add"
        type="button"
      >
        m+
      </button>
      <button
        className="btn small"
        data-name="msubtract"
        data-value=""
        title="memory subtract"
        type="button"
      >
        m−
      </button>
      <button
        className="btn small"
        data-name="mclear"
        title="memory clear"
        type="button"
      >
        mc
      </button>
      <button
        className="btn small"
        data-name="mrecall"
        title="memory recall"
        type="button"
      >
        mr
      </button>
      <button
        className="btn small"
        data-action="divide"
        data-name="division"
        data-value="÷"
        title="divide"
        type="button"
      >
        ÷
      </button>
    </div>
    <div
      className="numbers"
    >
      <button
        className="btn"
        data-name="number7"
        data-value="7"
        title="number 7"
        type="button"
      >
        7
      </button>
      <button
        className="btn"
        data-name="number8"
        data-value="8"
        title="number 8"
        type="button"
      >
        8
      </button>
      <button
        className="btn"
        data-name="number9"
        data-value="9"
        title="number 9"
        type="button"
      >
        9
      </button>
      <button
        className="btn"
        data-name="number4"
        data-value="4"
        title="number 4"
        type="button"
      >
        4
      </button>
      <button
        className="btn"
        data-name="number5"
        data-value="5"
        title="number 5"
        type="button"
      >
        5
      </button>
      <button
        className="btn"
        data-name="number6"
        data-value="6"
        title="number 6"
        type="button"
      >
        6
      </button>
      <button
        className="btn"
        data-name="number1"
        data-value="1"
        title="number 1"
        type="button"
      >
        1
      </button>
      <button
        className="btn"
        data-name="number2"
        data-value="2"
        title="number 2"
        type="button"
      >
        2
      </button>
      <button
        className="btn"
        data-name="number3"
        data-value="3"
        title="number 3"
        type="button"
      >
        3
      </button>
      <button
        className="btn"
        data-name="number0"
        data-value="0"
        title="number 0"
        type="button"
      >
        0
      </button>
      <button
        className="btn"
        data-name="dot"
        data-value="."
        title="dot"
        type="button"
      >
        .
      </button>
      <button
        className="btn"
        data-name="clear"
        title="clear"
        type="button"
      >
        c
      </button>
    </div>
    <div
      className="operators"
    >
      <button
        className="btn small"
        data-action="multiply"
        data-name="multiplication"
        data-value="×"
        title="multiply"
        type="button"
      >
        ×
      </button>
      <button
        className="btn small"
        data-action="subtract"
        data-name="subtraction"
        data-value="-"
        title="subtract"
        type="button"
      >
        −
      </button>
      <button
        className="btn small"
        data-action="sum"
        data-name="addition"
        data-value="+"
        title="sum"
        type="button"
      >
        +
      </button>
      <button
        className="btn small"
        data-name="equality"
        title="equals"
        type="button"
      >
        =
      </button>
    </div>
  </div>
</form>
`;

exports[`Macwidget should render without crashing 1`] = `
<form
  className="Macwidget"
>
  <output
    className="display"
  />
  <div
    className="keyboard"
  >
    <div
      className="brackets"
    >
      <button
        className="btn small"
        data-name="madd"
        data-value=""
        title="memory add"
        type="button"
      >
        m+
      </button>
      <button
        className="btn small"
        data-name="msubtract"
        data-value=""
        title="memory subtract"
        type="button"
      >
        m−
      </button>
      <button
        className="btn small"
        data-name="mclear"
        title="memory clear"
        type="button"
      >
        mc
      </button>
      <button
        className="btn small"
        data-name="mrecall"
        title="memory recall"
        type="button"
      >
        mr
      </button>
      <button
        className="btn small"
        data-action="divide"
        data-name="division"
        data-value="÷"
        title="divide"
        type="button"
      >
        ÷
      </button>
    </div>
    <div
      className="numbers"
    >
      <button
        className="btn"
        data-name="number7"
        data-value="7"
        title="number 7"
        type="button"
      >
        7
      </button>
      <button
        className="btn"
        data-name="number8"
        data-value="8"
        title="number 8"
        type="button"
      >
        8
      </button>
      <button
        className="btn"
        data-name="number9"
        data-value="9"
        title="number 9"
        type="button"
      >
        9
      </button>
      <button
        className="btn"
        data-name="number4"
        data-value="4"
        title="number 4"
        type="button"
      >
        4
      </button>
      <button
        className="btn"
        data-name="number5"
        data-value="5"
        title="number 5"
        type="button"
      >
        5
      </button>
      <button
        className="btn"
        data-name="number6"
        data-value="6"
        title="number 6"
        type="button"
      >
        6
      </button>
      <button
        className="btn"
        data-name="number1"
        data-value="1"
        title="number 1"
        type="button"
      >
        1
      </button>
      <button
        className="btn"
        data-name="number2"
        data-value="2"
        title="number 2"
        type="button"
      >
        2
      </button>
      <button
        className="btn"
        data-name="number3"
        data-value="3"
        title="number 3"
        type="button"
      >
        3
      </button>
      <button
        className="btn"
        data-name="number0"
        data-value="0"
        title="number 0"
        type="button"
      >
        0
      </button>
      <button
        className="btn"
        data-name="dot"
        data-value="."
        title="dot"
        type="button"
      >
        .
      </button>
      <button
        className="btn"
        data-name="clear"
        title="clear"
        type="button"
      >
        c
      </button>
    </div>
    <div
      className="operators"
    >
      <button
        className="btn small"
        data-action="multiply"
        data-name="multiplication"
        data-value="×"
        title="multiply"
        type="button"
      >
        ×
      </button>
      <button
        className="btn small"
        data-action="subtract"
        data-name="subtraction"
        data-value="-"
        title="subtract"
        type="button"
      >
        −
      </button>
      <button
        className="btn small"
        data-action="sum"
        data-name="addition"
        data-value="+"
        title="sum"
        type="button"
      >
        +
      </button>
      <button
        className="btn small"
        data-name="equality"
        title="equals"
        type="button"
      >
        =
      </button>
    </div>
  </div>
</form>
`;
